### ISSUE TITLE
Adding Non-Nullable Property to Discriminator Class Causes NullReferenceException in EntityLoadInfo.Materialize

### ISSUE THREAD

**Issue by DanielStout5** [dotnet/efcore#18346]
I have a database table that uses a discriminator like so in `OnModelCreating`:

            builder.Entity<Business>()
                .HasDiscriminator(x => x.Type)
                .HasValue<Shop>(BusinessType.Shop)
                .HasValue<Brand>(BusinessType.Brand);

I added a bool property to the `Shop` class. The migration had `nullable: true` specified in the `AlterColumn` statement, so it was added as nullable in the database. I assume this is necessary because the bool is only on one of the classes that are stored on the same table.

When running the application it throws this exception:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at lambda_method(Closure , MaterializationContext )
   at Microsoft.EntityFrameworkCore.Query.EntityLoadInfo.Materialize()
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.GetEntity(IKey key, EntityLoadInfo entityLoadInfo, Boolean queryStateManager, Boolean throwOnNullKey)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.BufferedEntityShaper`1.Shape(QueryContext queryContext, ValueBuffer& valueBuffer)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.BufferlessMoveNext(DbContext _, Boolean buffer, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
```

Changing the type to `bool?` makes it work fine, but I would have expected an error when attempting to add the migration, not a confusing runtime NullReferenceException.

### Further technical details

EF Core version: 2.2.3
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 2.2
Operating system: Windows 10
IDE: Visual Studio 2019 16.3.3

**Comment by smitpatel** [comment]
This is likely fixed in 3.0.

**Comment by ajcvickers** [comment]
@smitpatel Is there an issue tracking the fix in 3.0?

**Comment by smitpatel** [comment]
Not that I am aware of.
